lib/itertools/stream/stream.go

```diff
@@ -0,0 +1,459 @@
+				var zero T
+				yield(zero, trace.Wrap(err))
+				return
+			}
+
+			if n > 0 {
+				n--
+				continue
+			}
+
+			if !yield(item, nil) {
+				return
+			}
+		}
+	}
+}
+
+// Flatten flattens a stream of streams into a single stream of items.
+func Flatten[T any](stream Stream[Stream[T]]) Stream[T] {
+	return func(yield func(T, error) bool) {
+		for inner, err := range stream {
+			if err != nil {
+				var zero T
+				yield(zero, trace.Wrap(err))
+				return
+			}
+
+			for item, err := range inner {
+				if err != nil {
+					var zero T
+					yield(zero, trace.Wrap(err))
+					return
+				}
+
+				if !yield(item, nil) {
+					return
+				}
+			}
+		}
+	}
+}
+
+// MapErr maps over the "terminating" error value of a stream. This is a distinctly different
+// concept than mapping over the second value of an iter.Seq2. The mapping function is called with
+// a nil value when the stream terminates successfully, and nil values returned by the function are
+// not yielded. The effect of this behavior is that MapErr can be used to suppress, inject, *or* modify
+// the terminating error of a stream.
+func MapErr[T any](stream Stream[T], fn func(error) error) Stream[T] {
+	return func(yield func(T, error) bool) {
+		for item, err := range stream {
+			if err != nil {
+				mappedErr := fn(err)
+				if mappedErr == nil {
+					// terminating error was suppressed
+					return
+				}
+				var zero T
+				yield(zero, mappedErr)
+				return
+			}
+
+			if !yield(item, nil) {
+				return
+			}
+		}
+
+		if err := fn(nil); err != nil {
+			var zero T
+			yield(zero, trace.Wrap(err))
+		}
+	}
+}
+
+// RateLimit applies a rate-limiting function to a stream before each attempt to get the
+// next item. If the wait function returns a non-nil error, the stream is halted. The wait
+// function may return io.EOF to indicate a graceful/expected halting condition.
+func RateLimit[T any](stream Stream[T], wait func() error) Stream[T] {
+	return func(yield func(T, error) bool) {
+		for item, err := range stream {
+			if err != nil {
+				var zero T
+				yield(zero, trace.Wrap(err))
+				return
+			}
+
+			if !yield(item, nil) {
+				return
+			}
+
+			if err := wait(); err != nil {
+				if !errors.Is(err, io.EOF) {
+					var zero T
+					yield(zero, trace.Wrap(err))
+				}
+				return
+			}
+		}
+	}
+}
+
+// MergeStreams merges two sorted streams and returns a single stream which uses the provided less function to determine which item to yield first in order to preserve the sort order.
+func MergeStreams[T any](
+	streamA Stream[T],
+	streamB Stream[T],
+	less func(a, b T) bool,
+) Stream[T] {
+	return func(yield func(T, error) bool) {
+		var itemA, itemB T
+		var okA, okB bool
+		var err error
+
+		nextA, stopA := iter.Pull2(streamA)
+		nextB, stopB := iter.Pull2(streamB)
+		defer stopA()
+		defer stopB()
+
+		itemA, err, okA = nextA()
+		if err != nil {
+			var zero T
+			yield(zero, trace.Wrap(err))
+			return
+		}
+
+		itemB, err, okB = nextB()
+		if err != nil {
```

Comment:
Should we yield the other item before terminating with an error, or should we just drop it?